### PR TITLE
[SPARK-26772][YARN] Delete ServiceCredentialProvider and make HadoopDelegationTokenProvider a developer API

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

`HadoopDelegationTokenProvider` has basically the same functionality just like `ServiceCredentialProvider` so the interfaces can be merged.

`YARNHadoopDelegationTokenManager` now loads `ServiceCredentialProvider`s in one step. The drawback of this if one provider fails all others are not loaded. `HadoopDelegationTokenManager` loads `HadoopDelegationTokenProvider`s independently so it provides more robust behaviour.

In this PR I've I've made the following changes:
* Deleted `YARNHadoopDelegationTokenManager` and `ServiceCredentialProvider`
* Made `HadoopDelegationTokenProvider` a `@DeveloperApi`

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
cc @vanzin

**Comment by vanzin** [comment]
I'm starting to question whether we should keep this interface. It's in the yarn module, which means it's not considered public by Spark, which leads to issues (such as we don't check for compatibility, and it's been broken in the past). Pinging @mridulm for comments.

It might be a better idea to perhaps make the core interface a `@DeveloperApi` instead of keeping this one around and never really checking that it works properly across releases.

**Comment by gaborgsomogyi** [comment]
Presume @vanzin meant make `HadoopDelegationTokenProvider` a `@DeveloperApi`. In general I support this idea, only one question. [Here](https://issues.apache.org/jira/browse/SPARK-26254?focusedCommentId=16720241&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16720241) you've mentioned `ServiceLoader` had problems with non-private to Spark API's, won't be a problem here(don't know the context)?

@mridulm All in all I think it's good idea to move the custom provider load support into `HadoopDelegationTokenManager`. In order to do this [SPARK-26766](https://issues.apache.org/jira/browse/SPARK-26766) has to be resolved first(another reason why move `YarnSparkHadoopUtil.hadoopFSsToAccess`). What I not yet caught why `HadoopDelegationTokenManager` should be `@DeveloperApi`? Presume only `HadoopDelegationTokenProvider` has to be exposed, hasn't it?

**Comment by vanzin** [comment]
I don't think the manager needs to be exposed, just the provider interface. That's the same thing on the YARN side today.

BTW Gabor has already done all the rest of the work to make this possible, the only thing missing is actually making the interface public.

**Comment by mridulm** [comment]
Sorry, I meant `HadoopDelegationTokenProvider` and not `HadoopDelegationTokenManager` - copy paste error !

**Comment by vanzin** [comment]
So have we reached a consensus to drop this and expose the interface in core as a developer API?

**Comment by gaborgsomogyi** [comment]
From my perspective yes but until now not yet reached to continue this PR.

**Comment by vanzin** [comment]
I see that you implemented the "make the core interface public" approach and deleted the YARN version; but now the PR description and the bug don't really reflect the code anymore.

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenProvider.scala [hunk 1]
```diff
@@ -18,15 +18,16 @@
...
 import org.apache.hadoop.security.Credentials
 
 import org.apache.spark.SparkConf
+import org.apache.spark.annotation.DeveloperApi
 
 /**
  * Hadoop delegation token provider.
  */
-private[spark] trait HadoopDelegationTokenProvider {
+@DeveloperApi
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenProvider.scala]: You need to update the doc comment (see javadoc for `DeveloperApi`), because scaladoc is weird.

Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenProvider.scala]: Fixed.

**Comment by gaborgsomogyi** [comment]
> the PR description and the bug don't really reflect the code anymore.
Yeah, updated.

**Comment by gaborgsomogyi** [comment]
Retest this please

**Comment by vanzin** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenProvider.scala [hunk 2]
```diff
@@ -18,15 +18,17 @@
 package org.apache.spark.deploy.security
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenProvider.scala]: I remembered that things under `org.apache.spark.deploy` are not really "public" in Spark's view (i.e. they're not publicly documented). So leaving this interface here would mean it still remains in an awkward situation.

We should probably move the interface to `org.apache.spark.security` (the rest could stay here). Up to you if you want to do it here or in a separate patch.

Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenProvider.scala]: I think it's not a heavy stuff so put it here.

**Comment by vanzin** [comment]
Merging to master.
